manuals/fsclfsinfo.1
.Dd June 26, 2026
.Dt FSCLFSINFO 1
.Os
.Sh NAME
.Nm fsclfsinfo
.Nd determines information about a Common Log File System (CLFS) store
.Sh SYNOPSIS
.Nm fsclfsinfo
.Op Fl hvV
.Ar source
.Sh DESCRIPTION
.Nm fsclfsinfo
is a utility to determine information about a Common Log File System (CLFS) \
store
.Pp
.Nm fsclfsinfo
is part of the
.Nm libfsclfs
package.
.Nm libfsclfs
is a library to access the Common Log File System (CLFS) format
.Pp
.Ar source
is the source file.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl h
shows this help
.It Fl v
verbose output to stderr
.It Fl V
print version
.El
.Sh ENVIRONMENT
None
.Sh FILES
None
.Sh EXAMPLES
.Bd -literal
# fsclfsinfo /dev/sda1
fsclfsinfo 20130915
.sp
Common Log File System (CLFS) store information:
.sp
.Ed
.Sh DIAGNOSTICS
Errors, verbose and debug output are printed to stderr when verbose output \
\-v is enabled.
Verbose and debug output are only printed when enabled at compilation.
.Sh AUTHORS
.An Joachim Metz <joachim.metz@gmail.com>
.Sh BUGS
Please report bugs of any kind on the project issue tracker: \
https://github.com/libyal/libfsclfs/issues
.Sh COPYRIGHT
Copyright (C) 2010-2026, Joachim Metz <joachim.metz@gmail.com>.
.sp
This is free software; see the source for copying conditions.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A \
PARTICULAR PURPOSE.


ChangeLog
TODO
* don't allow store to be opened multiple times

* add check signature functions?
* use notify output grouping
* fsclfsinfo:
  - add info handle
* code clean up:
  - init functions
* owner page
  - detect last valid owner page
* add caches (use fdata)
  - blocks/records
  - owner pages

Format
* improve fixup handling (what about lower nibble of lower byte)
* add checksum handing

API
* store: add get stream by name function

Debug
* print offsets of containers
* libfusn: add debug descriptions of update resource and source flags

Deployment
* check configure test of sub library included functions
* write man pages

20141011
* see `git log' for more recent change log
* removed README.macosx
* updated dependencies
* changes for project site move

20140529
* updated dependencies

20140118
* 2014 update
* updated dependencies

20131109
* updated dependencies

20130915
* code clean up

20130914
* updated dependencies
* added libcthreads build support

20130224
* updated dependencies
* 2013 update

20130214
* worked on debug output

20120929
* updated dependencies

20120527
* updated dependencies
* code clean up:
  - extern definitions

20110313
* code clean up

20110305
* updated libfdata

20110304
* updated configure.ac, common, libcstring, liberror, libnotify, libsystem,
  libuna, libbfio, libfdatetime, libfwnt
* worked on gettext support
* worked on TxR support
* 2011 update

20101220
* removed libfdata

20101219
* updated common, libcstring, libsystem, libbfio, libfdatetime
* updated array type
* code clean up

20101203
* code clean up
* updated libfdata

20101202
* worked on format
  - owner page
  - virual/physical lsn mappings
  - block chains
  - split blocks

20101201
* worked on format
  - container
* worked on container and stream
* refactored base log into store
* updated libbfio

20101130
* worked on format
  - base log, block fixup values
* added container and stream

20101129
* worked on format
  - base log
* changed handle into store and base log

20101128
* worked on format

20101127
* updated libfdatetime
* moved abort to IO handle
* moved record to record value
* record is now an external 'object' type
* worked on format

20101126
* updated configure.ac
* updated libcstring, liberror, libuna, libbfio, libfdata
* added libclfs.rc
* worked on format

20101113
* textual changes
* worked on initial version
* worked on format

20101112
* created initial version
* worked on format



fsclfstools/fsclfsinfo.c
/*
 * Shows information obtained from a Common Log File System (CLFS) store.
 *
 * Copyright (C) 2010-2026, Joachim Metz <joachim.metz@gmail.com>
 *
 * Refer to AUTHORS for acknowledgements.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#include <common.h>
#include <file_stream.h>
#include <system_string.h>
#include <types.h>

#if defined( HAVE_FCNTL_H ) || defined( WINAPI )
#include <fcntl.h>
#endif

#if defined( HAVE_IO_H ) || defined( WINAPI )
#include <io.h>
#endif

#if defined( HAVE_STDLIB_H ) || defined( WINAPI )
#include <stdlib.h>
#endif

#if defined( HAVE_UNISTD_H )
#include <unistd.h>
#endif

#include "fsclfstools_getopt.h"
#include "fsclfstools_libcerror.h"
#include "fsclfstools_libclocale.h"
#include "fsclfstools_libcnotify.h"
#include "fsclfstools_libfsclfs.h"
#include "fsclfstools_output.h"
#include "fsclfstools_signal.h"
#include "fsclfstools_unused.h"
#include "info_handle.h"

info_handle_t *fsclfsinfo_info_handle = NULL;
int fsclfsinfo_abort                  = 0;

/* Signal handler for fsclfsinfo
 */
void fsclfsinfo_signal_handler(
      fsclfstools_signal_t signal FSCLFSTOOLS_ATTRIBUTE_UNUSED )
{
	libcerror_error_t *error = NULL;
	static char *function    = "fsclfsinfo_signal_handler";

	FSCLFSTOOLS_UNREFERENCED_PARAMETER( signal )

	fsclfsinfo_abort = 1;

	if( fsclfsinfo_info_handle != NULL )
	{
		if( info_handle_signal_abort(
		     fsclfsinfo_info_handle,
		     &error ) != 1 )
		{
			libcnotify_printf(
			 "%s: unable to signal info handle to abort.\n",
			 function );

			libcnotify_print_error_backtrace(
			 error );
			libcerror_error_free(
			 &error );
		}
	}
	/* Force stdin to close otherwise any function reading it will remain blocked
	 */
#if defined( WINAPI ) && !defined( __CYGWIN__ )
	if( _close(
	     0 ) != 0 )
#else
	if( close(
	     0 ) != 0 )
#endif
	{
		libcnotify_printf(
		 "%s: unable to close stdin.\n",
		 function );
	}
}

/* The main program
 */
#if defined( HAVE_WIDE_SYSTEM_CHARACTER )
int wmain( int argc, wchar_t * const argv[] )
#else
int main( int argc, char * const argv[] )
#endif
{
	const char *description = \
		"Use fsclfsinfo to determine information about a Common Log File System (CLFS) store.";

	fsclfstools_option_t options[ ] = {
		{ 'h', NULL, "shows this help" },
		{ 'v', NULL, "verbose output to stderr" },
		{ 'V', NULL, "print version" },
		{ 0, "source", "the source CLFS base log file" },
	};
	system_character_t options_string[ 32 ];

	libfsclfs_error_t *error   = NULL;
	system_character_t *source = NULL;
	char *program              = "fsclfsinfo";
	system_integer_t option    = 0;
	int number_of_options      = (int) ( sizeof( options ) / sizeof( fsclfstools_option_t ) );
	int verbose                = 0;

#if defined( __MINGW32__ ) && defined( HAVE_MINGW_BINMODE )
	_setmode( _fileno( stdout ), _O_BINARY );
	_setmode( _fileno( stderr ), _O_BINARY );
#endif

	libcnotify_stream_set(
	 stderr,
	 NULL );
	libcnotify_verbose_set(
	 1 );

	if( libclocale_initialize(
	     "fsclfstools",
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize locale values.\n" );

		goto on_error;
	}
	if( fsclfstools_output_initialize(
	     _IONBF,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize output settings.\n" );

		goto on_error;
	}
	fsclfstools_output_version_fprint(
	 stdout,
	 program );

	if( fsclfstools_getopt_get_options_string(
	     options,
	     number_of_options,
	     options_string,
	     32 ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to determine options string.\n" );

		goto on_error;
	}
	while( ( option = fsclfstools_getopt(
	                   argc,
	                   argv,
	                   options_string ) ) != (system_integer_t) -1 )
	{
		switch( option )
		{
			case (system_integer_t) '?':
			default:
				fprintf(
				 stderr,
				 "Invalid argument: %" PRIs_SYSTEM "\n",
				 argv[ optind - 1 ] );

				fsclfstools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				return( EXIT_FAILURE );

			case (system_integer_t) 'h':
				fsclfstools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				return( EXIT_SUCCESS );

			case (system_integer_t) 'v':
				verbose = 1;

				break;

			case (system_integer_t) 'V':
				fsclfstools_output_copyright_fprint(
				 stdout );

				return( EXIT_SUCCESS );
		}
	}
	if( optind == argc )
	{
		fprintf(
		 stderr,
		 "Missing source file.\n" );

		fsclfstools_getopt_usage_fprint(
		 stdout,
		 program,
		 description,
		 options,
		 number_of_options );

		return( EXIT_FAILURE );
	}
	source = argv[ optind ];

	libcnotify_verbose_set(
	 verbose );
	libfsclfs_notify_set_stream(
	 stderr,
	 NULL );
	libfsclfs_notify_set_verbose(
	 verbose );

	if( info_handle_initialize(
	     &fsclfsinfo_info_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize info handle.\n" );

		goto on_error;
	}
	if( info_handle_open_input(
	     fsclfsinfo_info_handle,
	     source,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to open source file.\n" );

		goto on_error;
	}
	if( info_handle_store_fprint(
	     fsclfsinfo_info_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to print store information.\n" );

		goto on_error;
	}
	if( info_handle_close_input(
	     fsclfsinfo_info_handle,
	     &error ) != 0 )
	{
		fprintf(
		 stderr,
		 "Unable to close info handle.\n" );

		goto on_error;
	}
	if( info_handle_free(
	     &fsclfsinfo_info_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to free info handle.\n" );

		goto on_error;
	}
	return( EXIT_SUCCESS );

on_error:
	if( error != NULL )
	{
		libcnotify_print_error_backtrace(
		 error );
		libcerror_error_free(
		 &error );
	}
	if( fsclfsinfo_info_handle != NULL )
	{
		info_handle_free(
		 &fsclfsinfo_info_handle,
		 NULL );
	}
	return( EXIT_FAILURE );
}

